module/zfs/zil.c

```diff
@@ -723,9 +725,39 @@ zil_create(zilog_t *zilog)
 	 * (zh is part of the MOS, so we cannot modify it in open context.)
 	 */
 	if (tx != NULL) {
+		/*
+		 * If "zilsaxattr" feature is enabled on zpool, then activate
+		 * it.
+		 */
+		if (spa_feature_is_enabled(zilog->zl_spa,
+		    SPA_FEATURE_ZILSAXATTR)) {
+			mutex_enter(&ds->ds_lock);
+			ds->ds_feature_activation[SPA_FEATURE_ZILSAXATTR] =
+			    (void *)B_TRUE;
+			mutex_exit(&ds->ds_lock);
+		}
 		dmu_tx_commit(tx);
 		txg_wait_synced(zilog->zl_dmu_pool, txg);
+	} else {
+		if (spa_feature_is_enabled(zilog->zl_spa,
+		    SPA_FEATURE_ZILSAXATTR) &&
+		    !dsl_dataset_feature_is_active(ds,
```

Comment:
Regarding our earlier (offline) question about when it's safe to check for feature enablement/activation status.

`spa_feature_is_enabled` is safe because it used the `spa_feat_refcount_cache`, which is updated atomically from `feature_sync`.
Technically, `spa_feature_is_enabled` returns `B_TRUE` a little early because `spa_feat_refcount_cache` is updated _before_ the txg that contains the feature enablement change has finished syncing out.
But oh well, any observer of the `B_TRUE` value can only affect changes in later txgs.
So everything is fine.

Now, regarding `spa_feature_is_active`: that uses the same `spa_feat_refcount_cache` infrastructure.
But some features can flip back from active to enabled.
So, observers that make changes based on `spa_feature_is_active == B_TRUE` must be certain that the feature will still be active when the change syncs out.

And finally, regarding `dsl_dataset_feature_is_active`: it seems to me like it's only safe to use that function from syncing context. The reason is that `ds_feature` updates are only made in syncing context and are not guarded by locks.
That being said, there are a bunch of places in the code base where we call `dsl_dataset_feature_is_active` outside of syncing context.
I guess that, while its technically a data race, there won't be any partial reads on Intel platforms, so we're safe.
But this is probably something that should be fixed.

Reply:
Current patch seems working :). So, would avoid making any more changes until necessary.

Reply:
As discussed, keeping it. Added comment explaining need of using dsl_dataset_feature_is_active();